src/lib/ctable.lua

```diff
@@ -217,6 +217,30 @@ function CTable:insert(hash, key, value, updates_allowed)
    return index
 end
 
+-- Choose a random index between the start of the table and its occupancy_hi
+-- value. This guarantees that there will be an entry to eject, because this
+-- is only called when the table is 'full'.
+local function random_eject(ctab)
+   local eject_index = math.random(0, ctab.occupancy_hi - 1)
+   -- Empty entries can't be ejected; find a non-empty one
+   while ctab.entries[eject_index].hash == HASH_MAX do
+      eject_index = eject_index + 1
+   end
+   assert(eject_index <= ctab.size + ctab.max_displacement,
+      "Ctab: eject_index too large!") -- This should be unreachable
+   local ptr = ctab.entries + eject_index
+   ctab:remove_ptr(ptr)
+end
+
+-- Behave exactly like insertion, except if the table is full: if it is, then
+-- eject a random entry instead of resizing.
+function CTable:add_with_random_ejection(key, value, updates_allowed)
```

Comment:
I did it this way because it's so reliant on ctable internals, but I'm ok with the wrapper class approach - it's just something that can easily be incidentally be broken by seemingly fine changes upstream until the wrapper class and appropriate tests are upstreamed.